File: user-impl/src/main/java/com/gabizou/cameraiq/demo/impl/UserServiceImpl.java
package com.gabizou.cameraiq.demo.impl;

import akka.NotUsed;
import akka.japi.Pair;
import com.gabizou.cameraiq.demo.api.User;
import com.gabizou.cameraiq.demo.api.UserId;
import com.gabizou.cameraiq.demo.api.UserRegistration;
import com.gabizou.cameraiq.demo.api.UserService;
import com.gabizou.cameraiq.demo.util.DemoFunctional;
import com.gabizou.cameraiq.demo.util.UUIDType5;
import com.google.inject.Inject;
import com.lightbend.lagom.javadsl.api.ServiceCall;
import com.lightbend.lagom.javadsl.persistence.PersistentEntityRef;
import com.lightbend.lagom.javadsl.persistence.PersistentEntityRegistry;
import org.apache.log4j.LogManager;
import org.apache.log4j.Logger;
import org.pcollections.POrderedSet;
import org.taymyr.lagom.javadsl.openapi.AbstractOpenAPIService;

import java.util.Set;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

public class UserServiceImpl extends AbstractOpenAPIService implements UserService {

    private static final String ENTITY_KEY = UserServiceImpl.class.getName();
    private final PersistentEntityRegistry registry;
    private final UserRepository repo;
    private static final Logger LOGGER = LogManager.getLogger("UserService");

    @Inject
    public UserServiceImpl(PersistentEntityRegistry registry,
                           UserRepository repository) {
        this.registry = registry;
        registry.register(UserEntity.class);
        this.repo = repository;
    }

    @Override
    public ServiceCall<NotUsed, POrderedSet<User>> getUsers() {
        return none -> this.repo.getUsers();
    }

    @Override
    public ServiceCall<Set<UserId>, Pair<POrderedSet<User>, Set<String>>> getUsersByIds() {
        return userIds -> {
            final POrderedSet<ServiceCall<NotUsed, User>> serviceCalls = userIds.parallelStream()
                .map(this::lookupUser)
                .collect(DemoFunctional.toImmutableSet());
            // Gather all futures to handle, with the exception handling below
            // in the accumilator of our custom collector
            final CompletableFuture<Void> allToProcess =
                CompletableFuture.allOf(serviceCalls.toArray(new CompletableFuture[serviceCalls.size()]));
            return allToProcess.thenApply(v -> serviceCalls.parallelStream()
                .map(ServiceCall::invoke)
                .map(CompletionStage::toCompletableFuture)
                .collect(DemoFunctional.toPairOfObjectAndExceptions()));


        };
    }

    private PersistentEntityRef<UserCommand> getUserRef() {
        return this.registry.refFor(UserEntity.class, UserServiceImpl.ENTITY_KEY);
    }

}

File: user-impl/src/main/scala/com/gabizou/demo/impl/UserServiceImpl.scala
package com.gabizou.demo.impl

import java.time.Instant

import akka.NotUsed
import akka.actor.Status.Success
import akka.stream.scaladsl.Source
import com.gabizou.cameraiq.demo.util.UUIDType5
import com.gabizou.demo.api.{User, UserId, UserRegistration, UserService}
import com.lightbend.lagom.scaladsl.api.ServiceCall
import com.lightbend.lagom.scaladsl.persistence.PersistentEntityRegistry
import org.apache.log4j.{LogManager, Logger}

import scala.concurrent.{ExecutionContext, Future}

class UserServiceImpl(registry: PersistentEntityRegistry, repo: UserRepository) extends UserService {
  override def lookupUser(userId: UserId): ServiceCall[NotUsed, User] = _ => {
    entityRef.ask(GetUser(userId))
      .flatMap(_ => {
        UserServiceImpl.LOGGER.info(s"Looking up user $userId")
        repo.lookupUser(userId)
      })(ExecutionContext.global)
  }

  override def createUser: ServiceCall[UserRegistration, User] = registrationInfo => {
    val uuid = UUIDType5.nameUUIDFromNamespaceAndString(UUIDType5.NAMESPACE_OID, registrationInfo.email)
    val user = User(UserId(uuid), registrationInfo)
    UserServiceImpl.LOGGER.info(s"Creating new user $registrationInfo")
    repo.saveUser(user)
        .flatMap(saved => {
          entityRef.ask(CreateUser(saved, Instant.now()))
        })(ExecutionContext.global)
  }

  override def getUsers: ServiceCall[NotUsed, Set[User]] = _ => repo.getUsers

  override def getUsersByIds: ServiceCall[Source[UserId, NotUsed], Source[User, NotUsed]] = users => {
    val source = users.watchTermination() { (_, done) =>
      done.onComplete {
        case Success => UserSErviceImpl.LOG
      }

    }
  }

  private def entityRef = registry.refFor[UserEntity](UserServiceImpl.ENTITY_KEY)
}
object UserServiceImpl {
  private val LOGGER: Logger = LogManager.getLogger("UserService")
  private val ENTITY_KEY = classOf[UserServiceImpl].getName
}
